pkg/openfeature/event_executor_test.go

```diff
@@ -0,0 +1,1307 @@
+		t.Fatal(err)
+	}
+
+	// api level handlers
+	executor.registerApiHandler(ProviderReady, &errorCallback)
+	executor.registerApiHandler(ProviderReady, &successAPICallback)
+
+	// provider association
+	providerName := "providerA"
+
+	// client level handlers
+	executor.registerClientHandler(providerName, ProviderReady, &errorCallback)
+	executor.registerClientHandler(providerName, ProviderReady, &successClientCallback)
+
+	// trigger events manually
+	go func() {
+		eventing.Invoke(Event{
+			ProviderName:         providerName,
+			EventType:            ProviderReady,
+			ProviderEventDetails: ProviderEventDetails{},
+		})
+	}()
+
+	select {
+	case <-rsp:
+		break
+	case <-time.After(200 * time.Millisecond):
+		t.Error("API level callback timeout - executor recovery was not successful")
+	}
+
+	select {
+	case <-rspClient:
+		break
+	case <-time.After(200 * time.Millisecond):
+		t.Error("client callback timeout - executor recovery was not successful")
+	}
+}
+
+// Requirement 5.3.1 If the provider's initialize function terminates normally, PROVIDER_READY handlers MUST run.
+func TestEventHandler_InitOfProvider(t *testing.T) {
+	t.Run("for default provider in global handler scope", func(t *testing.T) {
+		defer t.Cleanup(initSingleton)
+
+		// provider
+		provider := struct {
+			FeatureProvider
+			StateHandler
+		}{
+			NoopProvider{},
+			&stateHandlerForTests{
+				State: NotReadyState,
+			},
+		}
+
+		// callback
+		rsp := make(chan EventDetails, 1)
+		callback := func(e EventDetails) {
+			rsp <- e
+		}
+
+		AddHandler(ProviderReady, &callback)
+		err := SetProvider(provider)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		select {
+		case <-rsp:
+			break
+		case <-time.After(200 * time.Millisecond):
+			t.Errorf("timedout waiting for ready state callback, but got none")
+		}
+	})
+
+	t.Run("for default provider with unassociated client handler", func(t *testing.T) {
+		defer t.Cleanup(initSingleton)
+
+		// provider
+		provider := struct {
+			FeatureProvider
+			StateHandler
+		}{
+			NoopProvider{},
+			&stateHandlerForTests{
+				State: NotReadyState,
+			},
+		}
+
+		// callback
+		rsp := make(chan EventDetails, 1)
+		callback := func(e EventDetails) {
+			rsp <- e
+		}
+
+		addClientHandler("clientWithNoProviderAssociation", ProviderReady, &callback)
+		err := SetProvider(provider)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		select {
+		case <-rsp:
+			break
+		case <-time.After(200 * time.Millisecond):
+			t.Errorf("timedout waiting for ready state callback, but got none")
+		}
+	})
+
+	t.Run("for named provider in client scope", func(t *testing.T) {
+		defer t.Cleanup(initSingleton)
+
+		// provider
+		provider := struct {
+			FeatureProvider
+			StateHandler
+		}{
+			NoopProvider{},
+			&stateHandlerForTests{
+				State: NotReadyState,
+			},
+		}
+
+		// callback
+		rsp := make(chan EventDetails, 1)
+		callback := func(e EventDetails) {
+			rsp <- e
+		}
+
+		addClientHandler("someClient", ProviderReady, &callback)
```

Comment:
These new tests are very comprehensive, they add a lot of confidence. I've noticed that with the client tests, we're always testing the private client methods. Do you think it might be better to extract the client tests into `client_test.go` and instead of calling the private `addClientHandler`, call the public equivalent on the client? Would that cause a lot of difficulty?

Reply:
Good suggestion. I tried to aggregate all eventing spec-bound tests [1] under `event_executor_test.go`. But yes, using internal methods is not the best choice here. 

I will do the following - given that we have client API level test coverage [2], I will avoid internal methods and migrate this registration to be done via a Client. This way we validate both client API as well as eventing functionality 

[1] - https://github.com/open-feature/spec/blob/main/specification/sections/05-events.md
[2] - https://github.com/open-feature/go-sdk/pull/190/files#diff-1520c2162931fb780e1961a5cd73b7a724cbbbecea232fccb9384db38f4957c7R92-R101

Reply:
@toddbaert  thank you for highlighting this. 

I migrated tests to use API contracts where possible. For example, I am using internals only where it's neede (ex:- to check if registrations are correct or validations are there). However, I kept tests in the same file. This is to aggregate all eventing spec tests at one location

Reply:
> @toddbaert thank you for highlighting this.
> 
> I migrated tests to use API contracts where possible. For example, I am using internals only where it's neede (ex:- to check if registrations are correct or validations are there). However, I kept tests in the same file. This is to aggregate all eventing spec tests at one location

That sounds perfect to me.